controllers/rollout/trafficrouting.go

```diff
@@ -0,0 +1,283 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rollout
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	appsv1alpha1 "github.com/openkruise/rollouts/api/v1alpha1"
+	"github.com/openkruise/rollouts/controllers/rollout/trafficrouting"
+	"github.com/openkruise/rollouts/controllers/rollout/trafficrouting/nginx"
+	"github.com/openkruise/rollouts/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	utilpointer "k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *rolloutContext) doCanaryTrafficRouting() (bool, error) {
+	canaryStatus := r.newStatus.CanaryStatus
+	//fetch stable service
+	sName := r.rollout.Spec.Strategy.Canary.TrafficRouting.Service
+	r.stableService = &corev1.Service{}
+	err := r.Get(context.TODO(), client.ObjectKey{Namespace: r.rollout.Namespace, Name: sName}, r.stableService)
+	if err != nil {
+		klog.Errorf("rollout(%s/%s) get stable service(%s) failed: %s", r.rollout.Namespace, r.rollout.Name, sName, err.Error())
+		// not found, wait a moment, retry
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	canaryStatus.CanaryService = fmt.Sprintf("%s-canary", sName)
+	// fetch canary service
+	// todo for the time being, we do not consider the scenario where the user only changes the stable service definition during rollout progressing
+	r.canaryService = &corev1.Service{}
+	err = r.Get(context.TODO(), client.ObjectKey{Namespace: r.rollout.Namespace, Name: canaryStatus.CanaryService}, r.canaryService)
+	if err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("rollout(%s/%s) get canary service(%s) failed: %s", r.rollout.Namespace, r.rollout.Name, canaryStatus.CanaryService, err.Error())
+		return false, err
+	} else if errors.IsNotFound(err) {
+		klog.Infof("rollout(%s/%s) canary service(%s) Not Found, and create it", r.rollout.Namespace, r.rollout.Name, canaryStatus.CanaryService)
+		if err = r.createCanaryService(); err != nil {
+			return false, err
+		}
+		by, _ := json.Marshal(r.canaryService)
+		klog.Infof("create rollout(%s/%s) canary service(%s) success", r.rollout.Namespace, r.rollout.Name, string(by))
+	}
+
+	// update service selector
+	// update service selector specific revision pods
+	if r.canaryService.Spec.Selector[r.podRevisionLabelKey()] != canaryStatus.CanaryRevision {
+		r.canaryService.Spec.Selector[r.podRevisionLabelKey()] = canaryStatus.CanaryRevision
+		if err = r.retryUpdateService(r.canaryService); err != nil {
+			return false, err
+		}
+		// update canary service time, and wait 3 seconds, just to be safe
```

Comment:
what are we waiting for,  can we check the condition explicitly? The same suggestion apply to the following waits